### PR TITLE
[WIP] Adding first draft of recommendations for reusable code

### LINKED ISSUE #1596: N simple rules for making your code reusable?
### Summary

I had to give a short talk on "how to make your code reusable" for my university (Aalto University, Finland) and wrote this down for the occasion:

Text: https://hackmd.io/@AaltoSciComp/howToMakeYourCodeReusable
Video of the talk: https://www.youtube.com/watch?v=FdMEeUqhJNw&feature=youtu.be

They are a collection of "N simple rules" for reaching multiple levels of code re-usability (from re-running of everything as it was when the original results where created, to more generalisable code that will live its own life). I start from the "reproducibility matrix" and re-adapt those concepts to software re-usability.

My question for you is: do you think this is valuable for the Turing way? It is a text that basically points to many different parts of the Turing way book as well as other hands-on teaching materials (e.g. CodeRefinery), so eventually a section like this could live at the end of the guide for reproducible research, ... but I am unsure if it adds really anything or if readers would find it useful, so I would like to have some fresh independent eyes to check it and give an opinion. 

Side note: I am also happy to get contributors/co-authors if you get inspired by the idea and want to help.

### What needs to be done?

[ ] Here an opinion from core Turing way members who curated the guide to reproducibility part, to see if there is any value in the text/lecture I linked above

### Who can help?

* I am not sure, sorry

---

### Updates

Updates will be here

### PR DESCRIPTION
**Pull request by c-martinez**
### Summary

Fixes #1596

This PR compiles recommendations for reusable research from multiple sources. 

### List of changes proposed in this PR (pull-request)

* Pointer from the `Open Source` chapter to the specific recommendations for code reuse
* New chapter with recommendations for code reuse (styled as a checklist)

### What should a reviewer concentrate their feedback on?

- [ ] It should be clear why code re-use is important
- [ ] It should be clear there are different types of re-use use cases (and when each one is applicable)
- [ ] It should be clear which recommendations apply for which type of re-use

### Acknowledging contributors

@eglerean, @avakiai, @bast, @annefou, @fdiblen all contributed in different ways.

- [ ] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/main/README.md#contributors) in the README file.
- [ ] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/main/README.md#contributors) in the README file:

### CODE REVIEW

File: book/website/reproducible-research/code-reuse/code-reuse.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+(rr-code-reuse)=
+# Reusable code
+Your software project could range from a small script you use for data processing, a notebook used for data analysis, or a software library implementing your algorithms. Regardless of how big or small your software project is, it is important to make your code reusable.
```
Comment (fdiblen) [inline book/website/reproducible-research/code-reuse/code-reuse.md:3]: range from .... to ...?

File: book/website/reproducible-research/code-reuse/code-reuse.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+This section contains a checklist of recommendations for making your software more reusable. The {ref}`rr-code-reuse-recommendation-details` section contains a more in-depth explanation of each of these recommendations. You can follow the recommendations that are more suitable for your type of software and skip the ones which are not relevant in your case.
+
+### Repeatable recommendations
+1. Make sure you can find it (in space)
+1. Make sure you can find it (in time)
+1. Make sure you can execute the same sequence of operations
+1. Make sure your environment and sequence of operations is robust and no human is needed to replicate what was done
+1. License your code
+  - with a license that allows for reuse;
+  - with a license compatible with the dependencies’ licenses
```
Comment (abelsiqueira) [inline book/website/reproducible-research/code-reuse/code-reuse.md:36]: The indentation is broken
image
Maybe the following:
```suggestion
1. License your code
    - with a license that allows for reuse;
    - with a license compatible with the dependencies’ licenses
```

File: book/website/reproducible-research/code-reuse/code-reuse.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+1. Make sure it is citatable
```
Comment (abelsiqueira) [inline book/website/reproducible-research/code-reuse/code-reuse.md:37]:
```suggestion
1. Make sure it is citable
```

**Review by abelsiqueira — COMMENTED** [review]
I noticed some details, hope they are helpful.

File: book/website/reproducible-research/code-reuse/code-reuse.md (hunk 1 above)
Comment (c-martinez) [reply book/website/reproducible-research/code-reuse/code-reuse.md:3]: Good points both.

File: book/website/_bibliography/references.bib [hunk 4]
```diff
@@ -1001,12 +1001,36 @@ @article{fong2017manipulation
...
+
+@Article{ChueHong2021FAIR4RS,
+  author    = {Chue Hong, Neil P. and Katz, Daniel S. and Barker, Michelle and Lamprecht, Anna-Lena and Martinez, Carlos and Psomopoulos, Fotis E. and Harrow, Jen and Castro, Leyla Jael and Gruenpeter, Morane and Martinez, Paula Andrea and Honeyman, Tom},
+  title     = {FAIR Principles for Research Software (FAIR4RS Principles)},
+  doi       = {10.15497/RDA00065},
+  language  = {en},
+  copyright = {Creative Commons Attribution 4.0 International},
+  publisher = {Research Data Alliance},
+  year      = {2021},
+}
```
[the remark this answers is not in the document]
Comment (c-martinez) [reply book/website/_bibliography/references.bib:1036]: It is published as an RDA output. I'll leave journal empty for now.

File: book/website/reproducible-research/code-reuse/code-reuse.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+## How to make your code more reusable
+This section contains a checklist of recommendations for making your software more reusable. The {ref}`rr-code-reuse-recommendation-details` section contains a more in-depth explanation of each of these recommendations. You can follow the recommendations that are more suitable for your type of software and skip the ones which are not relevant in your case.
+
+### Repeatable recommendations
+1. Make sure you can find it (in space)
+1. Make sure you can find it (in time)
+1. Make sure you can execute the same sequence of operations
+1. Make sure your environment and sequence of operations is robust and no human is needed to replicate what was done
+1. License your code
```
Comment (annefou) [inline book/website/reproducible-research/code-reuse/code-reuse.md:39]: Is there any plan to add links to help choosing the license (for instance https://choosealicense.com/)?

Comment (c-martinez) [reply book/website/reproducible-research/code-reuse/code-reuse.md:39]: More or less: it should point to the relevant section -- in this case, https://the-turing-way.netlify.app/reproducible-research/licensing/licensing-software.html, which in turn includes a link to  https://choosealicense.com/.

**Comment by c-martinez** [comment]
@fdiblen, @annefou, @eglerean, @abelsiqueira -- I've made a few changes You can view the [Netlify preview here](https://deploy-preview-2182--the-turing-way.netlify.app/reproducible-research/code-reuse/code-reuse.html#rr-code-reuse)

What do you think? I would like to merge this sooner rather than later (and then iterate on it if needed).

**Comment by bast** [comment]
Thanks a lot for this! I have minor change suggestions but they can be done as follow-up PRs and don't have to delay this issue.

And there is a problem with the numbering (many items are numbered as 1.) but also that can be fixed as follow-up.

**Comment by c-martinez** [comment]
Thanks @bast -- I've fixed the numbering, and if others also approve (or at least do not object) we can merge it and improve on it in the future :-)

**Review by abelsiqueira — APPROVED** [review]
Looks good to me :cake:

**Comment by fdiblen** [comment]
@c-martinez Looks good to me too. Thank you for the work!

**Review by fdiblen — APPROVED** [review]

**Pull request merged**